Detectors/PHOS/reconstruction/src/Clusterer.cxx

```diff
@@ -23,87 +26,140 @@ using namespace o2::phos;
+    mClusters.clear(); // internal list of FullClusters
 
-  LOG(DEBUG) << "Number of PHOS clusters" << clusters->size();
+    if (!mBadMap) {
+      if (o2::phos::PHOSSimParams::Instance().mCCDBPath.compare("localtest") == 0) {
+        mBadMap = new BadChannelMap(1);    // test default map
+        mCalibParams = new CalibParams(1); //test calibration map
+        LOG(INFO) << "No reading BadMap/Calibration from ccdb requested, set default";
+      } else {
+        LOG(INFO) << "Getting BadMap object from ccdb";
+        o2::ccdb::CcdbApi ccdb;
+        std::map<std::string, std::string> metadata; // do we want to store any meta data?
+        ccdb.init("http://ccdb-test.cern.ch:8080");  // or http://localhost:8080 for a local installation
+        long bcTime = 1;                             //TODO!!! Convert BC time to time o2::InteractionRecord bcTime = digitsTR.front().getBCData() ;
+        mBadMap = ccdb.retrieveFromTFileAny<o2::phos::BadChannelMap>("PHOS/BadMap", metadata, bcTime);
+        mCalibParams = ccdb.retrieveFromTFileAny<o2::phos::CalibParams>("PHOS/Calib", metadata, bcTime);
+        if (!mBadMap) {
+          LOG(FATAL) << "[PHOSCellConverter - run] can not get Bad Map";
+        }
+        if (!mCalibParams) {
+          LOG(FATAL) << "[PHOSCellConverter - run] can not get CalibParams";
+        }
+      }
+    }
 
-  // Unfold overlapped clusters
-  // TODO:
-  // MakeUnfolding();
+    // Collect digits to clusters
+    makeClusters(digits);
 
-  // Calculate properties of collected clusters (Local position, energy, disp etc.)
-  EvalCluProperties(digits, clusters);
+    LOG(DEBUG) << "Number of PHOS clusters" << clusters->size();
+    // Unfold overlapped clusters
+    // Split clusters with several local maxima if necessary
+    if (o2::phos::PHOSSimParams::Instance().mUnfoldClusters) {
+      makeUnfoldings(digits);
+    }
+    // Calculate properties of collected clusters (Local position, energy, disp etc.)
+    evalCluProperties(digits, clusters, dmc, cluMC);
+
+    trigRec->emplace_back(tr.getBCData(), indexStart, clusters->size());
+  }
   LOG(DEBUG) << "PHOS clustrization done";
 }
 //____________________________________________________________________________
-void Clusterer::MakeClusters(const std::vector<Digit>* digits, std::vector<Cluster>* clusters)
+void Clusterer::makeClusters(const std::vector<Digit>* digits)
 {
   // A cluster is defined as a list of neighbour digits
-  const double kClusteringThreshold = 0.050; // TODO: To be read from RecoParam
-  const double kDigitMinEnergy = 0.010;      // TODO: to be implemented as a digit energy cut
 
   // Mark all digits as unused yet
-  const int maxNDigits = 3584; // There is no clusters larger than PHOS module ;)
-  int nDigits = digits->size();
+  const int maxNDigits = 12546; // There is no digits more than in PHOS modules ;)
   bool digitsUsed[maxNDigits];
 
-  for (int i = 0; i < nDigits; i++) {
+  for (int i = 0; i < maxNDigits; i++) {
```

Comment:
instead of the init loop it is much more efficient to  ``memset(digitsUsed,0,sizeof(bool)*maxNDigits)``

Reply:
done